### PR TITLE
Game components

### PR DESCRIPTION
**Pull request by HourGlss**
players can move, have skills, health, mana. map is buildable.

need to work on exporting map, exporting players, exporting mobs, exporting rooms.

not sure if you wanna pull now or later?

### CODE REVIEW

**Review by BautistaDalmasso — APPROVED** [review]
LGTM, although I would have appreciated more comments in the code :P .

File: server/game_components/game_objects.py [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+        self.__min_damage = _min_damage
+        self.__max_damage = _max_damage
+        self.__cost = _cost
+        self.__hit_percentage = _hit_percentage
+        self.name = _name
+
+    def action(self, _caster: Entity, _target: Entity | None) -> list[dict]:
+        if _target is None:
+            assert self.requires_target is False
+        """
```
Comment (AbooMinister25) [inline server/game_components/game_objects.py:140]: The docstring should preferably be at the top of the function definition.

File: server/game_components/game_objects.py [hunk 2]
```diff
@@ -0,0 +1,533 @@
...
+                                "hit": hit,
+                                "dmg": dmg,
+                                "cast": cast,
+                            }
+                        )
+                return action_list
+            else:
+                _target = _caster
+
+        if isinstance(_target, Entity) and (not self.area_of_effect):
```
Comment (AbooMinister25) [inline server/game_components/game_objects.py:210]: The parenthesis around `not self.area_of_effect` is unnecessary, you can omit them.

File: server/game_components/game_objects.py [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+            dmg = random.randint(self.__min_damage, self.__max_damage)
+            if cast:
+                hit_check = random.randint(0, 100)
+                if hit_check <= self.__hit_percentage:
```
Comment (AbooMinister25) [inline server/game_components/game_objects.py:214]: You can shorted this to `hit = hit_check <= self.__hit_percentage`

**Pull request merged**